soc/arm/st_stm32/stm32g0/power.c

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (c) 2021 STMicroelectronics.
+ * Copyright (c) 2021 G-Technologies Sdn. Bhd.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <zephyr.h>
+#include <pm/pm.h>
+#include <soc.h>
+#include <init.h>
+
+#include <stm32g0xx_ll_utils.h>
+#include <stm32g0xx_ll_bus.h>
+#include <stm32g0xx_ll_cortex.h>
+#include <stm32g0xx_ll_pwr.h>
+#include <stm32g0xx_ll_rcc.h>
+#include <stm32g0xx_ll_system.h>
+#include <clock_control/clock_stm32_ll_common.h>
+
+#include <logging/log.h>
+LOG_MODULE_DECLARE(soc, CONFIG_SOC_LOG_LEVEL);
+
+/* Invoke Low Power/System Off specific Tasks */
+void pm_power_state_set(struct pm_state_info info)
+{
+	if (info.state != PM_STATE_SUSPEND_TO_IDLE) {
+		LOG_DBG("Unsupported power state %u", info.state);
+		return;
+	}
+
+	switch (info.substate_id) {
+	case 1: /* this corresponds to the STOP0 mode: */
+		/* enter STOP0 mode */
+		LL_PWR_SetPowerMode(LL_PWR_MODE_STOP0);
+		LL_LPM_EnableDeepSleep();
+		/* enter SLEEP mode : WFE or WFI */
+		k_cpu_idle();
+		break;
+	case 2: /* this corresponds to the STOP1 mode: */
+		/* enter STOP1 mode */
+		LL_PWR_SetPowerMode(LL_PWR_MODE_STOP1);
+		LL_LPM_EnableDeepSleep();
+		/* enter SLEEP mode : WFE or WFI */
+		k_cpu_idle();
+		break;
+	default:
+		LOG_DBG("Unsupported power state substate-id %u",
+			info.substate_id);
+		break;
+	}
+}
+
+/* Handle SOC specific activity after Low Power Mode Exit */
+void pm_power_state_exit_post_ops(struct pm_state_info info)
+{
+	if (info.state != PM_STATE_SUSPEND_TO_IDLE) {
+		LOG_DBG("Unsupported power substate-id %u", info.state);
+	} else {
+		switch (info.substate_id) {
+		case 1:	/* STOP0 */
```

Comment:
I think we should be consistent with the behavior of other series and reset the system control SLEEPONEXIT bit using LL_LPM_DisableSleepOnExit().

@FRASTM Currently I don't see how we would land here after a wake-up from a sleep-on-exit ISR. 
For which circumstances do we need to reset this bit here?

Reply:
>  reset the system control SLEEPONEXIT bit using LL_LPM_DisableSleepOnExit()

Sure thing, after I have some time to test this.

>I don't see how we would land here after a wake-up from a sleep-on-exit ISR

Not sure if I get you right: [1](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/pm/power.c#L160) -> [2](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/pm/power.c#L163) -> [3](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/pm/power.c#L138)

Reply:
> > reset the system control SLEEPONEXIT bit using LL_LPM_DisableSleepOnExit()
> 
> Sure thing, after I have some time to test this.

@str4t0m Based on my test log, `LL_LPM_DisableSleepOnExit()` doesn't have the same effect as `LL_LPM_EnableSleep()`. The device will still go into stop mode when Zephyr enters tickless idle thread unintentionally.

You can use the sample below to test with, it is based on the `power_mgmt/blinky`:
[not_blinky.txt](https://github.com/zephyrproject-rtos/zephyr/files/6761172/not_blinky.txt)

It it is sleeping and waking properly, you will see these in the terminal:
```
Next 5 seconds should be sleeping
Next 5 seconds will not be sleeping
Next 5 seconds should be sleeping
Next 5 seconds will not be sleeping
```

and with a multimeter you can observe that the current switch between tickless idle & stop mode. On my custom board it's 6.x mA & 0.5x mA

But with `LL_LPM_DisableSleepOnExit()`, it would print half of them as if the baudrate is wrong, because it entered stop mode unintentionally as the idle thread runs and nothing restore the RCC when it exits. The current consumption will always be at stop mode level since now it would be switching between Zephyr's stop mode and unintentional stop mode.

If you build the same file with `CONFIG_SHELL=y`, you will notice that you wont be able to type anything into the console anymore once it entered stop mode, also because of unconfigure RCC due to the unintentional stop mode.